File: python/rayz/cube.py
from __future__ import annotations

import math

from rayz.constants import EPSILON
from rayz.intersection import Intersection
from rayz.shape import Shape
from rayz.tuple import Vector


class Cube(Shape):
    def local_intersect(self, ray) -> list:
        xtmin, xtmax = _check_axis(ray.origin.x, ray.direction.x)
        ytmin, ytmax = _check_axis(ray.origin.y, ray.direction.y)
        ztmin, ztmax = _check_axis(ray.origin.z, ray.direction.z)

        tmin = max(xtmin, ytmin, ztmin)
        tmax = min(xtmax, ytmax, ztmax)

        if tmin > tmax:
            return []
        return [Intersection(tmin, self), Intersection(tmax, self)]

    def local_normal_at(self, point, hit=None) -> Vector:
        ax, ay, az = abs(point.x), abs(point.y), abs(point.z)
        maxc = max(ax, ay, az)
        if maxc == ax:
            return Vector(point.x, 0, 0)
        elif maxc == ay:
            return Vector(0, point.y, 0)
        return Vector(0, 0, point.z)

    def bounds(self):
        from rayz.bounds import Bounds
        from rayz.tuple import Point

        return Bounds(Point(-1, -1, -1), Point(1, 1, 1))


def _check_axis(origin: float, direction: float) -> tuple[float, float]:
    tmin_num = -1 - origin
    tmax_num = 1 - origin
    if abs(direction) >= EPSILON:
        tmin = tmin_num / direction
        tmax = tmax_num / direction
    else:
        # Ray is parallel to this axis's planes. A numerator of exactly
        # zero means the origin sits exactly on that plane, which imposes
        # no constraint from this axis rather than the NaN that
        # `0 * inf` would otherwise produce.
        tmin = -math.inf if tmin_num == 0 else tmin_num * math.inf
        tmax = math.inf if tmax_num == 0 else tmax_num * math.inf
    if tmin > tmax:
        return tmax, tmin
    return tmin, tmax

File: jruby/lib/rayz/cube.rb
require "matrix"
require_relative "shape"

module Rayz
  class Cube < Shape
    def local_intersect(local_ray)
      xtmin, xtmax = check_axis(local_ray.origin.x, local_ray.direction.x)
      ytmin, ytmax = check_axis(local_ray.origin.y, local_ray.direction.y)
      ztmin, ztmax = check_axis(local_ray.origin.z, local_ray.direction.z)

      tmin = [xtmin, ytmin, ztmin].max
      tmax = [xtmax, ytmax, ztmax].min

      return [] if tmin > tmax

      [
        Intersection.new(t: tmin, object: self),
        Intersection.new(t: tmax, object: self)
      ]
    end

    def local_normal_at(local_point, hit = nil)
      maxc = [local_point.x.abs, local_point.y.abs, local_point.z.abs].max

      if maxc == local_point.x.abs
        Vector.new(x: local_point.x, y: 0, z: 0)
      elsif maxc == local_point.y.abs
        Vector.new(x: 0, y: local_point.y, z: 0)
      else
        Vector.new(x: 0, y: 0, z: local_point.z)
      end
    end

    def bounds
      # Unit cube extends from -1 to 1 in all dimensions
      Bounds.new(
        min: Point.new(x: -1, y: -1, z: -1),
        max: Point.new(x: 1, y: 1, z: 1)
      )
    end

    private

    def check_axis(origin, direction)
      tmin_numerator = -1 - origin
      tmax_numerator = 1 - origin

      if direction.abs >= Util::EPSILON
        tmin = tmin_numerator / direction
        tmax = tmax_numerator / direction
      else
        tmin = tmin_numerator * Float::INFINITY
        tmax = tmax_numerator * Float::INFINITY
      end

      if tmin > tmax
        [tmax, tmin]
      else
        [tmin, tmax]
      end
    end
  end
end
